packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -563,15 +583,49 @@ export class ApolloServerBase {
         )
       ).filter(
         (maybeServerListener): maybeServerListener is GraphQLServerListener =>
-          typeof maybeServerListener === 'object' &&
-          !!maybeServerListener.serverWillStop,
+          typeof maybeServerListener === 'object',
       );
-      this.toDispose.add(async () => {
-        await Promise.all(
-          serverListeners.map(({ serverWillStop }) => serverWillStop?.()),
-        );
+
+      serverListeners.forEach(({ schemaDidChange, serverWillStop }) => {
+        let unsubscribe: Unsubscriber | undefined;
+        if (schemaDidChange) {
+          if (maybeGateway) {
+            // Note that latestApiSchema should have been set at least once by now, specifically
+            // during the call to gateway.load().
+            if (!latestApiSchema) {
+              throw new Error(
+                `Gateway did not notify listeners when loading schema.`,
+              );
+            }
+            // Note that it's important for there to be no await between this schemaDidChange()
+            // invocation and the registration of the new schema change callback, otherwise plugins
+            // could miss schema changes.
+            schemaDidChange({
+              apiSchema: latestApiSchema,
+              coreSchemaSdl: latestCoreSchemaSdl,
+            });
+            unsubscribe = maybeGateway.onSchemaChange(
+              (apiSchema, coreSchemaSdl) => {
+                schemaDidChange({ apiSchema, coreSchemaSdl });
```

Comment:
It seems like this needs to be better synchronized with the logic in the existing onSchemaChange. eg, we shouldn't call it if we don't actually update the schema because we're stopping or something, and we shouldn't call it until after the schema is actually updated (ie, schemaDerivedData is successfully changed).  Probably this means we need to save the handlers somewhere and call them from the existing onSchemaChange.

Reply:
Agree here, I've changed `_start()` to add listeners to a set and the gateway listener created in `startGatewayAndLoadSchema()` to notify listeners in the set after the schema derived data is updated.